src/power_system_simulation/graph_processing.py

```diff
@@ -0,0 +1,191 @@
+            """
+            try:
+                nx.find_cycle(self.graph, sc)
+                return True
+            except NetworkXNoCycle:
+                return False
+
+        if has_cycle(self, source_vertex_id) is True:
+            raise GraphCycleError("graph should not have cycles")
+
+        
+
+    
+
+
+    def find_downstream_vertices(self, edge_id: int) -> List[int]:
+        """
+        Given an edge id, return all the vertices which are in the downstream of the edge,
+            with respect to the source vertex.
+            Including the downstream vertex of the edge itself!
+
+        Only enabled edges should be taken into account in the analysis.
+        If the given edge_id is a disabled edge, it should return empty list.
+        If the given edge_id does not exist, it should raise IDNotFoundError.
+
+
+        For example, given the following graph (all edges enabled):
+
+            vertex_0 (source) --edge_1-- vertex_2 --edge_3-- vertex_4
+
+        Call find_downstream_vertices with edge_id=1 will return [2, 4]
+        Call find_downstream_vertices with edge_id=3 will return [4]
+
+        Args:
+            edge_id: edge id to be searched
+
+        Returns:
+            A list of all downstream vertices.
+        """
+        # put your implementation here
+        pass
+
+    def find_alternative_edges(self, disabled_edge_id: int) -> List[int]:
+        #To find edges in graph
+        edge_found = False
+        edge_to_disable = None
+        for u, v, data in self.graph.edges(data=True):
+            if data.get('edge_id') == disabled_edge_id:
+                edge_found = True
+                edge_to_disable = (u, v)
+                break
+        
+        if not edge_found:
+            raise IDNotFoundError(f"Edge {disabled_edge_id} not found")
+        
+        # Create a temporary graph without the edge to disable
+        temp_G = self.graph.copy()
+        temp_G.remove_edge(*edge_to_disable)
+        
+        # Find the two components created by removing the edge
+        component_A = nx.node_connected_component(temp_G, edge_to_disable[0])
+        component_B = nx.node_connected_component(temp_G, edge_to_disable[1])
+        
+        # Check all disabled edges (not in current graph) that could connect A and B
+        alternatives = []
+        for u, v, data in self.graph.edges(data=True):
+            if not temp_G.has_edge(u, v):  # This edge is disabled
+                if (u in component_A and v in component_B) or (u in component_B and v in component_A):
+                    temp_G.add_edge(u, v)
+                    try:
+                        nx.find_cycle(temp_G)
+                        temp_G.remove_edge(u, v)
+                    except NetworkXNoCycle:
+                        # No cycle so this is a valid alternative
+                        alternatives.append(data['edge_id'])
+                        temp_G.remove_edge(u, v)
+        
```

Comment:
suggestion: It is a good practice to have your python code have as little nested statements as possible. This is a way to unnest the two if statements
```suggestion
            if temp_G.has_edge(u, v):  # This edge is enabled
                continue
                
            if (u in component_A and v in component_B) or (u in component_B and v in component_A):
                temp_G.add_edge(u, v)
                try:
                    nx.find_cycle(temp_G)
                    temp_G.remove_edge(u, v)
                except NetworkXNoCycle:
                        # No cycle so this is a valid alternative
                        alternatives.append(data['edge_id'])
                        temp_G.remove_edge(u, v)
        
```